src/main/utility/synchronization/thread_parking.rs

```diff
@@ -0,0 +1,104 @@
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::sync::Arc;
+
+/// Used to unpark a thread, but which hasn't been assigned a specific thread yet.
+#[derive(Debug, Clone)]
+pub struct ThreadUnparkerUnassigned {
+    ready_flag: Arc<AtomicBool>,
+}
+
+/// Used to unpark a thread.
+#[derive(Debug, Clone)]
+pub struct ThreadUnparker {
+    thread: std::thread::Thread,
+    ready_flag: Arc<AtomicBool>,
+}
+
+/// Used to park a thread. The `ThreadParker` is derived from a `ThreadUnparker` or
+/// `ThreadUnparkerUnassigned`, and must only be used on the thread which the unparker was assigned
+/// to. If the `ThreadUnparker` was assigned to thread A, then `ThreadParker::park()` must only be
+/// called from thread A.
+#[derive(Debug, Clone)]
+pub struct ThreadParker {
+    ready_flag: Arc<AtomicBool>,
+}
+
+impl ThreadUnparkerUnassigned {
+    pub fn new() -> Self {
+        Self {
+            ready_flag: Arc::new(AtomicBool::new(false)),
+        }
+    }
+
+    /// Assign this to a thread that will be unparked.
+    #[must_use]
+    pub fn assign(self, thread: std::thread::Thread) -> ThreadUnparker {
+        ThreadUnparker::new(self.ready_flag, thread)
+    }
+
+    /// Get a new [`ThreadParker`]. The `ThreadParker` must only be used from the thread which we
+    /// will later assign ourselves to using `assign()`. This is useful if you want to pass a
+    /// `ThreadParker` to a new thread before you have a handle to that thread.
+    pub fn parker(&self) -> ThreadParker {
+        ThreadParker::new(Arc::clone(&self.ready_flag))
+    }
+}
+
+impl ThreadUnparker {
+    fn new(ready_flag: Arc<AtomicBool>, thread: std::thread::Thread) -> Self {
+        Self { ready_flag, thread }
+    }
+
+    /// Unpark the assigned thread.
+    pub fn unpark(&self) {
+        self.ready_flag.store(true, Ordering::Release);
+        self.thread.unpark();
+    }
+
+    /// Get a new [`ThreadParker`] for the assigned thread.
+    pub fn parker(&self) -> ThreadParker {
+        ThreadParker::new(Arc::clone(&self.ready_flag))
+    }
+}
+
+impl ThreadParker {
+    fn new(ready_flag: Arc<AtomicBool>) -> Self {
+        Self { ready_flag }
+    }
+
+    /// Park the current thread until [`ThreadUnparker::unpark()`] is called. You must only call
+    /// `park()` from the thread which the corresponding `ThreadUnparker` is assigned, otherwise a
+    /// deadlock may occur.
+    pub fn park(&self) {
+        while self
+            .ready_flag
+            .compare_exchange(true, false, Ordering::SeqCst, Ordering::Relaxed)
+            .is_err()
+        {
```

Comment:
I think this can deadlock?

```
thread A: l72 park
thread A: l73 compare_exchange
thread B:     l53 unpark
thread B:     l54 ready_flag.store
thread B:     l55 thread.unpark
thread A: l78 thread::park
```

I'm not sure there's a way to fix it without a semaphore.

If the goal is primarily to allow the thread-pools etc to run under miri, maybe an alternative is for your custom semaphore type to delegate to Rust's std semaphore instead of the pthread semaphore when `cfg(miri)`?

Reply:
> I think this can deadlock?

I wouldn't be surprised if it does :) But I don't see the deadlock yet.

In your example, I think the `ready_flag` starts as false. The `compare_exchange()` in thread A fails to change `ready_flag` from true to false, then thread B changes `ready_flag` to true and calls `unpark()`. Thread A then calls `park()` which will return immediately because of the previous `unpark()`, loops back to the `compare_exchange()` which should now succeed since `ready_flag` is true, and we exit the loop.

> If the goal is primarily to allow the thread-pools etc to run under miri, maybe an alternative is for your custom semaphore type to delegate to Rust's std semaphore instead of the pthread semaphore when `cfg(miri)`?

There is no semaphore in the standard library unfortunately.

Reply:
> Thread A then calls park() which will return immediately because of the previous unpark()

Ahh, that's the part I was missing. I was thinking of `park` and `unpark` like a `FUTEX_WAIT` and `FUTEX_WAKE`, but it looks like it actually behaves more like a semaphore, which is what we want here.

Reply:
Yeah, I've been thinking of park and unpark like a binary semaphore. I'll add a comment about this.

The [documentation](https://doc.rust-lang.org/nightly/std/thread/fn.park.html) says:

> The `unpark` method on a `Thread` atomically makes the token available if it wasn’t already. Because the token is initially absent, `unpark` followed by `park` will result in the second call returning immediately.